src/nbl/video/IGPUDescriptorSet.cpp

```diff
@@ -5,81 +5,64 @@
-        break;
-    case asset::EDT_UNIFORM_BUFFER_DYNAMIC:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UBO_SSBOStorage.get()) + (m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER]);
-        break;
-    case asset::EDT_STORAGE_BUFFER_DYNAMIC:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_UBO_SSBOStorage.get()) + (m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_pool->m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC]);
-        break;
-    case asset::EDT_INPUT_ATTACHMENT:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_storageImageStorage.get()) + m_pool->m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE];
-        break;
-    case asset::EDT_ACCELERATION_STRUCTURE:
-        baseAddress = reinterpret_cast<core::smart_refctd_ptr<asset::IDescriptor>*>(m_pool->m_accelerationStructureStorage.get());
-        break;
-    default:
-        assert(!"Invalid code path.");
-        return nullptr;
-    }
+        const bool allowsFreeing = m_pool->m_creationParameters.flags.hasFlags(IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT);
+        for (auto i = 0u; i < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++i)
+        {
+            // There is no descriptor of such type in the set.
+            if (m_descriptorStorageOffsets.data[i] == ~0u)
+                continue;
 
-    return baseAddress;
-}
+            const auto type = static_cast<asset::IDescriptor::E_TYPE>(i);
 
-core::smart_refctd_ptr<IGPUSampler>* IGPUDescriptorSet::getMutableSamplerStorage() const
-{
-    return reinterpret_cast<core::smart_refctd_ptr<IGPUSampler>*>(m_pool->m_mutableSamplerStorage.get());
+            const uint32_t allocatedOffset = getDescriptorStorageOffset(type);
+            assert(allocatedOffset != ~0u);
+
+            const uint32_t count = m_layout->getTotalDescriptorCount(type);
+            assert(count != 0u);
+
+            std::destroy_n(m_pool->getDescriptorStorage(type) + allocatedOffset, count);
+
+            if (allowsFreeing)
+                m_pool->m_descriptorAllocators[i]->free(allocatedOffset, count);
+        }
+
+        const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
+        if (mutableSamplerCount > 0)
+        {
+            const uint32_t allocatedOffset = getMutableSamplerStorageOffset();
+            assert(allocatedOffset != ~0u);
+
+            std::destroy_n(m_pool->getMutableSamplerStorage() + allocatedOffset, mutableSamplerCount);
+
+            if (allowsFreeing)
+                m_pool->m_descriptorAllocators[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)]->free(allocatedOffset, mutableSamplerCount);
+        }
```

Comment:
probably best to wrap this all up into a `deleteSetStorage` (both free address and `std::destroy_n`) as per https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/8fbc87079edc5109a9c916eeb7ea7c0eff9ce128..b3719ddb1377d8af8f8d5f4dd112d3c844394553#r1116368522
```cpp
if (!isZombie())
    deleteSetStorage(this);
```

in there you could check if the Parent Pool allows freeing and decide whether to also deallocate the offsets after calling `std::destroy_n`

(furthermore if it doesn't allow freeing you could check if the number of allocated DS has dropped to 0 and if yes, call `reset_impl()` implicitly)

Reply:
Resolved, but I don't get what you mean by:
> (furthermore if it doesn't allow freeing you could check if the number of allocated DS has dropped to 0 and if yes, call reset_impl() implicitly)

Reply:
See  https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/1cd208bb98bdaa9f3eaef5530458ec23aeb2861b..ba8162877f6f8d193a0cfb77463d037fa260624e#r1121561950